flink-clients/src/main/java/org/apache/flink/client/program/AsyncClusterClientProvider.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.flink.client.program;
+
+import org.apache.flink.annotation.Internal;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/** Factory for {@link ClusterClient ClusterClients}. */
+@Internal
+public interface AsyncClusterClientProvider<T> extends ClusterClientProvider<T> {
```

Comment:
Given how this interface is used, I think we don't need it. We can also implement a `ClusterClientProvider` that resolves the application report in a blocking fashion.

Reply:
Removed it. Just introduce the class of `YarnClusterClientProvider` which is the implementation of `ClusterClientProvider`.

Right?